systems/ros/ros_subscriber_system.h

```diff
@@ -0,0 +1,235 @@
+  }
+
+  /**
+   * Returns the message counter stored in @p context.
+   */
+  int GetMessageCount(const drake::systems::Context<double>& context) const {
+    // Gets the last message count from abstract state.
+    return context.get_abstract_state<int>(kStateIndexMessageCount);
+  }
+
+ protected:
+  void DoCalcNextUpdateTime(const drake::systems::Context<double>& context,
+      drake::systems::CompositeEventCollection<double>* events, double* time)
+      const override {
+    // We do not support events other than our own message timing events.
+    LeafSystem<double>::DoCalcNextUpdateTime(context, events, time);
+    DRAKE_THROW_UNLESS(events->HasEvents() == false);
+    DRAKE_THROW_UNLESS(std::isinf(*time));
+
+    // Do nothing unless we have a new message.
+    const int last_message_count = GetMessageCount(context);
+    const int received_message_count = GetInternalMessageCount();
+    if (last_message_count == received_message_count) {
+      return;
+    }
+
+
+    *time = context.get_time();
+
+    drake::systems::EventCollection<
+        drake::systems::UnrestrictedUpdateEvent<double>>&
+        uu_events = events->get_mutable_unrestricted_update_events();
+    uu_events.add_event(
+        std::make_unique<drake::systems::UnrestrictedUpdateEvent<double>>(
+            drake::systems::Event<double>::TriggerType::kTimed));
+  }
+
+  void DoCalcUnrestrictedUpdate(
+      const drake::systems::Context<double>&, const std::vector<
+          const drake::systems::UnrestrictedUpdateEvent<double>*>&,
+      drake::systems::State<double>* state) const override {
+    ProcessMessageAndStoreToAbstractState(&state->get_mutable_abstract_state());
+  }
+
+  std::unique_ptr<drake::systems::AbstractValues> AllocateAbstractState()
+      const override {
+    std::vector<std::unique_ptr<drake::AbstractValue>> abstract_vals(2);
+    abstract_vals[kStateIndexMessage] =
+        this->RosSubscriberSystem::AllocateOutputValue();
+    abstract_vals[kStateIndexMessageCount] =
+        drake::AbstractValue::Make<int>(0);
+    return std::make_unique<drake::systems::AbstractValues>(
+        std::move(abstract_vals));
+  }
+
+  void SetDefaultState(const drake::systems::Context<double>& context,
+                       drake::systems::State<double>* state) const override {
+    ProcessMessageAndStoreToAbstractState(&state->get_mutable_abstract_state());
+  };
+
+ private:
+  void ProcessMessageAndStoreToAbstractState(
+      drake::systems::AbstractValues* abstract_state) const {
+    std::lock_guard<std::mutex> lock(received_message_mutex_);
+    abstract_state->get_mutable_value(kStateIndexMessage)
+        .get_mutable_value<RosMessage>() = received_message_;
+    abstract_state->get_mutable_value(kStateIndexMessageCount)
+        .get_mutable_value<int>() = received_message_count_;
+  };
+
+  // Callback entry point from ROS into this class. Also wakes up one thread
+  // block on notification_ if it's not nullptr.
```

Comment:
Not clear what this condition on "not being nullptr" is as there's no apparent conditional logic in the function implementation.

Reply:
Not sure what that comment means either. Deleted it.